src/libexpr/primops/fromYAML.cc

```diff
@@ -0,0 +1,156 @@
+
+            for (ryml::ConstNodeRef child : t.children()) {
+                std::string_view key(child.key().begin(), child.key().size());
+                visitYAMLNode(context, attrs.alloc(key), child);
+            }
+
+            v.mkAttrs(attrs);
+        }
+    } else if (t.is_seq()) {
+        context.state.mkList(v, t.num_children());
+
+        size_t i = 0;
+        for (ryml::ConstNodeRef child : t.children())
+            visitYAMLNode(context, *(v.listElems()[i++] = context.state.allocValue()), child);
+    } else if (valTypeCheck(ryml::TAG_NULL) && t.val_is_null()) {
+        v.mkNull();
+    } else if (t.has_val()) {
+        bool _bool;
+        NixFloat _float;
+        NixInt _int;
+        bool isQuoted = t.is_val_quoted();
+        auto val = t.val();
+        // Caution: ryml is able to convert integers into booleans and ryml::from_chars might ignore trailing chars
+        if (t.has_val_tag() && t.val_tag() == "!" && val.empty() && !isQuoted) { // special case for YAML string "!"
+            v.mkNull();
+        } else if (valTypeCheck(ryml::TAG_INT, !isQuoted) && val.is_integer() && ryml::from_chars(val, &_int)) {
+            v.mkInt(_int);
+        } else if (valTypeCheck(ryml::TAG_BOOL, !isQuoted) && ryml::from_chars(val, &_bool)) {
+            v.mkBool(_bool);
+        } else if (valTypeCheck(ryml::TAG_FLOAT, !isQuoted) && val.is_number() && ryml::from_chars(val, &_float)) {
+            v.mkFloat(_float);
+        }
+        if (valTypeCheck(ryml::TAG_STR) && v.type() == nThunk) {
+            std::string_view value(val.begin(), val.size());
+            v.mkString(value);
+        }
+    }
+    if (v.type() == nThunk) {
+        auto val = t.has_val() ? t.val() : "";
+        auto tag = t.has_val_tag() ? t.val_tag() : "";
+        auto msg = ryml::formatrs<std::string>(
+            "Error: The YAML value '{}' with '{}' tag cannot be represented as Nix data type", val, tag);
+        s_error(msg.data(), msg.size(), {}, &context);
+    }
+}
+
+static RegisterPrimOp primop_fromYAML({
+    .name = "__fromYAML",
+    .args = {"e"},
+    .doc = R"(
+      Convert a YAML 1.2 string to a Nix value, if a conversion is possible. For example,
+
+      ```nix
+      builtins.fromYAML ''{x: [1, 2, 3], y: !!str null, z: null}''
+      ```
+
+      returns the value `{ x = [ 1 2 3 ]; y = "null"; z = null; }`.
+
+      Maps are converted to attribute sets, but attribute sets require String keys, so that no other key data types are supported.
+
+      Scalars are converted to the type specified by their optional value tag and parsing fails, if a conversion is not possible.
+      Not all YAML types are supported by Nix, e.g. Nix has no binary and timestamp data types, so that parsing of YAML with any of these types fails.
+      Custom tags are ignored and a stream with multiple documents is mapped to a list except when the stream contains a single document.
+    )",
```

Comment:
This should be an experimental primop initially to gather user feedback, i.e.:

```suggestion
    )",
    .experimentalFeature = Xp::FromYaml,
```